### PR TITLE
chore(rust): update to latest master including quic-quinn PR

### PR DESCRIPTION
**Pull request by mxinden**
Updates to latest libp2p master plus
https://github.com/libp2p/rust-libp2p/pull/3454/, thus allowing us to test hole punch effectiveness of https://github.com/libp2p/rust-libp2p/pull/3454/.

I am behind a carrier grade NAT right now, thus hard to confirm this is working, thus draft for now.

### CODE REVIEW

**Comment by mxinden** [comment]
I don't have access to a punchable NAT right now. Do you @thomaseizinger? If so, mind running this version of the Rust punchr to test https://github.com/libp2p/rust-libp2p/pull/3454?

File: rust-client/src/main.rs [hunk 1]
```diff
@@ -4,21 +4,22 @@ use futures::stream::StreamExt;
 use libp2p::core::multiaddr::{Multiaddr, Protocol};
 use libp2p::core::muxing::StreamMuxerBox;
 use libp2p::core::transport::OrTransport;
-use libp2p::core::{upgrade, ConnectedPoint, ProtocolName, UpgradeInfo};
+use libp2p::core::{upgrade, ConnectedPoint};
 use libp2p::dns::DnsConfig;
+use libp2p::identify::Info;
```
Comment (thomaseizinger) [inline rust-client/src/main.rs:9]: Nit: Prefer `identify::Info` in the code.

File: rust-client/src/main.rs [hunk 2]
```diff
@@ -99,26 +100,14 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-                .map(|info| String::from_utf8(info.protocol_name().to_vec()))
-                .collect::<Result<Vec<String>, _>>()?;
-            let _ = protocols.insert(supported_protocols);
-        }
 
         let request = tonic::Request::new(grpc::RegisterRequest {
             client_id: local_peer_id.to_bytes(),
             agent_version: agent_version(),
-            protocols: protocols.clone().unwrap(),
+            // TODO: Send the right set of protocols.
```
Comment (thomaseizinger) [inline rust-client/src/main.rs:109]: Interesting, I guess we don't expose this at the moment. Perhaps this should be written as a `NetworkBehaviour`?

File: rust-client/src/main.rs [hunk 3]
```diff
@@ -190,19 +179,15 @@ async fn init_swarm(
 ) -> Result<Swarm<Behaviour>, Box<dyn std::error::Error>> {
     let local_peer_id = PeerId::from(local_key.public());
 
-    let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
-        .into_authentic(&local_key)
-        .expect("Signing libp2p-noise static DH keypair failed.");
```
Comment (thomaseizinger) [inline rust-client/src/main.rs:195]: 😍

File: rust-client/src/main.rs [hunk 4]
```diff
@@ -639,13 +641,11 @@ impl HolePunchAttemptState {
...
     bytes[0] = secret_key_seed;
-    let secret_key = identity::ed25519::SecretKey::from_bytes(&mut bytes)
-        .expect("this returns `Err` only if the length is wrong; the length is correct; qed");
-    identity::Keypair::Ed25519(secret_key.into())
+    identity::Keypair::ed25519_from_bytes(&mut bytes).unwrap()
 }
 
 #[derive(NetworkBehaviour)]
-#[behaviour(out_event = "Event", event_process = false)]
+#[behaviour(to_swarm = "Event", event_process = false)]
```
Comment (thomaseizinger) [inline rust-client/src/main.rs:648]:
```suggestion
#[behaviour(to_swarm = "Event")]
```

**Review by thomaseizinger — COMMENTED** [review]
Some thoughts :)

**Comment by thomaseizinger** [comment]
> I don't have access to a punchable NAT right now. Do you @thomaseizinger? If so, mind running this version of the Rust punchr to test https://github.com/libp2p/rust-libp2p/pull/3454?

I might have next week, not at the moment unfortunately.

**Comment by mxinden** [comment]
Results from a cafe:

image

**Comment by thomaseizinger** [comment]
> Results from a cafe:
> 
> image
> 

All QUIC?

**Comment by mxinden** [comment]
Good question. As far as I know not visible through the Grafana dashboard. Ran locally once more. The 4 last successful hole punches were all `/quic-v1`.

File: rust-client/src/main.rs (hunk 2 above)
Comment (mxinden) [reply rust-client/src/main.rs:109]: Possible. I suggest to delay it. I don't think the `protocols` field is needed as of today.

**Comment by mxinden** [comment]
I suggest we merge here only after https://github.com/libp2p/rust-libp2p/pull/3454.

**Comment by mxinden** [comment]
Still facing trouble with the Nix build. I don't foresee us using the Rust punchr client much in the future, especially not by external folks. Thus I am not going to proceed here. I don't think fixing the Nix issues is worth the time at this point.

**Comment by mxinden** [comment]
Closing here due to staleness.

**Pull request closed without merging**